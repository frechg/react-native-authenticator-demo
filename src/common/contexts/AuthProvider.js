import React from 'react'
import * as SecureStore from 'expo-secure-store';
import { initialAuthState, authStateReducer } from '../authStateReducer';
import * as Api from '../../services/auth';

export const AuthContext = React.createContext();

export const AuthProvider = (props) => {
  const [authState, authDispatch] = React.useReducer(authStateReducer, initialAuthState);

  const getAuthState = async () => {
    try {
      const token = await SecureStore.getItemAsync('authToken');
      const username = await SecureStore.getItemAsync('username');
      authDispatch({type: 'RESTORE_TOKEN', token: token, username: username});
    } catch (error) {
      throw new Error(error);
    }
  };

  const signIn = async (data) => {
    await Api.signIn(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        throw new Error('Email or password are incorrect.');
      }
    })
  };

  const signUp = async (data) => {
    await Api.signUp(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        const jsonResponse = await response.json();
        throw new Error(jsonResponse.errors[0])
      }
    });
  };

  const signOut = async () => {
    try {
      await SecureStore.deleteItemAsync('authToken');
      await SecureStore.deleteItemAsync('username');
      authDispatch({type: 'SIGN_OUT'});
    } catch (error) {
      throw new Error(error);
    }
  };

  const contextValue = React.useMemo(() => {
    return {authState, getAuthState, signUp, signIn, signOut};
  }, [authState]);

  return (
    <AuthContext.Provider value={contextValue}>
      {props.children}
    </AuthContext.Provider>
  );
};