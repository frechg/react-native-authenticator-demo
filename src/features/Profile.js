import React from 'react';
import { View, Text, Button, StatusBar } from 'react-native';
import { styles } from '../common/styles';
import { AuthContext } from '../common/contexts/AuthProvider';

export const Profile = () => {
  const { authState, signOut } = React.useContext(AuthContext);

  return (
    <View style={styles.container}>
      <Text>Welcome {authState.username}</Text>
      <Button title='Sign Out' onPress={() => signOut()} />
      <StatusBar style="auto" />
    </View>
  );
}