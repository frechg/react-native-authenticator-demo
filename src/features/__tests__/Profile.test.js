import * as React from 'react';
import { fireEvent, render, screen } from '@testing-library/react-native';

import { AuthContext } from '../../common/contexts/AuthProvider';
import { Profile } from '../Profile';

describe('<Profile />', () => {
  test('Renders Profile screen', () => {
    const signOut = jest.fn();   
    const authState = {username: 'Person'};
 
    render(
      <AuthContext.Provider value={{authState, signOut}}>
        <Profile />
      </AuthContext.Provider>
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('User can sign out', () => {
    const signOut = jest.fn();   
    const authState = {username: 'Person'};
 
    render(
      <AuthContext.Provider value={{authState, signOut}}>
        <Profile />
      </AuthContext.Provider>
    );

    fireEvent.press(screen.getByText('Sign Out'));

    expect(signOut).toHaveBeenCalled();
  });
});