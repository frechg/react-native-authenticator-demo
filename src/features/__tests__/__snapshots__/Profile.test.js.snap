// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> Renders Profile screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Welcome 
    Person
  </Text>
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Sign Out
      </Text>
    </View>
  </View>
</View>
`;
